File: backend-python/app/modules/system/service/config_service.py
from decimal import Decimal, InvalidOperation

from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.common.errors import BusinessError, bad_request, not_found
from app.common.id import new_id
from app.common.pagination import PageQuery, PageResult
from app.modules.system.dto.config_request import ConfigSaveRequest
from app.modules.system.entity import SysConfig


class ConfigService:
    def __init__(self, db: AsyncSession) -> None:
        self.db = db

    async def list_configs(
        self,
        page: PageQuery,
        config_name: str | None,
        config_key: str | None,
        status: int | None,
    ) -> PageResult[dict]:
        filters = [SysConfig.deleted == 0]
        if config_name and config_name.strip():
            filters.append(SysConfig.config_name.like(f"%{config_name.strip()}%"))
        if config_key and config_key.strip():
            filters.append(SysConfig.config_key.like(f"%{config_key.strip()}%"))
        if status is not None:
            filters.append(SysConfig.status == status)
        total = await self.db.scalar(select(func.count()).select_from(SysConfig).where(*filters)) or 0
        rows = (
            await self.db.scalars(
                select(SysConfig)
                .where(*filters)
                .order_by(SysConfig.created_at.desc(), SysConfig.id.desc())
                .offset((page.page - 1) * page.page_size)
                .limit(page.page_size)
            )
        ).all()
        return PageResult(
            list=[self._vo(row) for row in rows],
            total=str(total),
            page=page.page,
            pageSize=page.page_size,
        )

    async def public_configs(self) -> dict[str, str]:
        return {
            "systemName": await self.required_config("system.name"),
            "companyFullName": await self.config_or_default("system.company.fullName", ""),
            "logoUrl": await self.config_or_default("system.logo", ""),
            "watermarkEnabled": await self.config_or_default("system.watermark.enabled", "false"),
            "silentPrintEnabled": await self.config_or_default("print.silent.enabled", "false"),
        }

    async def save(self, config_id: int | None, request: ConfigSaveRequest) -> str | None:
        name = request.config_name.strip()
        key = request.config_key.strip()
        self._validate_value(request.value_type, request.config_value)
        duplicate_query = select(SysConfig.id).where(SysConfig.config_key == key, SysConfig.deleted == 0)
        if config_id is not None:
            duplicate_query = duplicate_query.where(SysConfig.id != config_id)
        if await self.db.scalar(duplicate_query) is not None:
            raise BusinessError(409000, "configKey already exists")

        if config_id is None:
            row = SysConfig(
                id=new_id(),
                config_name=name,
                config_key=key,
                config_value=request.config_value,
                value_type=request.value_type,
                builtin=0,
                status=1 if request.status is None else request.status,
                remark=_optional(request.remark),
                deleted=0,
            )
            self.db.add(row)
            await self.db.commit()
            return str(row.id)

        row = await self._find(config_id)
        row.config_name = name
        row.config_key = key
        row.config_value = request.config_value
        row.value_type = request.value_type
        row.remark = _optional(request.remark)
        if row.builtin != 1:
            row.status = 1 if request.status is None else request.status
        await self.db.commit()
        return None

    async def delete(self, config_id: int) -> None:
        row = await self._find(config_id)
        if row.builtin == 1:
            raise bad_request("operation failed")
        row.deleted = 1
        await self.db.commit()

    async def update_status(self, config_id: int, status: int | None) -> None:
        row = await self._find(config_id)
        if row.builtin == 1:
            raise bad_request("operation failed")
        row.status = 1 if status is None else status
        await self.db.commit()

    async def required_config(self, key: str) -> str:
        value = await self._config_value(key)
        if value is None or not value.strip():
            raise BusinessError(500000, f"system config missing: {key}")
        return value

    async def required_int_config(self, key: str) -> int:
        value = await self.required_config(key)
        try:
            return int(value)
        except ValueError as exc:
            raise BusinessError(500000, f"system config invalid: {key}") from exc

    async def config_or_default(self, key: str, default: str) -> str:
        value = await self._config_value(key)
        return default if value is None else value

    async def _config_value(self, key: str) -> str | None:
        return await self.db.scalar(
            select(SysConfig.config_value).where(
                SysConfig.config_key == key,
                SysConfig.deleted == 0,
                (SysConfig.builtin == 1) | (SysConfig.status == 1),
            )
        )

    async def _find(self, config_id: int) -> SysConfig:
        row = await self.db.scalar(
            select(SysConfig).where(SysConfig.id == config_id, SysConfig.deleted == 0)
        )
        if row is None:
            raise not_found()
        return row

    @staticmethod
    def _validate_value(value_type: str, value: str) -> None:
        if value_type == "boolean" and value.lower() not in {"true", "false"}:
            raise bad_request("configValue must be true or false")
        if value_type == "number":
            try:
                Decimal(value)
            except InvalidOperation as exc:
                raise bad_request("configValue must be number") from exc

    @staticmethod
    def _vo(row: SysConfig) -> dict:
        return {
            "id": str(row.id),
            "configName": row.config_name,
            "configKey": row.config_key,
            "configValue": row.config_value,
            "valueType": row.value_type,
            "builtin": row.builtin,
            "status": row.status,
            "remark": row.remark,
            "createdAt": row.created_at,
            "updatedAt": row.updated_at,
        }


def _optional(value: str | None) -> str | None:
    return value.strip() if value and value.strip() else None

File: backend-rust/src/modules/system/service/config_service.rs
use crate::common::{AppError, I64String, PageResult, next_id};
use crate::modules::system::dto::config_request::{ConfigQuery, ConfigSaveRequest};
use crate::modules::system::entity::sys_config::SysConfig;
use rbatis::RBatis;
use serde::Deserialize;
use serde_json::{Value, json};
use std::collections::BTreeMap;
use std::sync::Arc;

#[derive(Debug, Deserialize)]
struct CountRow {
    total: i64,
}

pub async fn public_config(database: Option<&Arc<RBatis>>) -> Result<Value, AppError> {
    let database = require_database(database)?;
    let rows: Vec<SysConfig> = database
        .exec_decode(
            "select id, config_name, config_key, config_value, value_type, builtin, status, remark, deleted from sys_config where deleted = 0 and config_key in (?, ?, ?, ?, ?) and (builtin = 1 or status = 1)",
            vec![
                rbs::value!("system.name"),
                rbs::value!("system.company.fullName"),
                rbs::value!("system.logo"),
                rbs::value!("system.watermark.enabled"),
                rbs::value!("print.silent.enabled"),
            ],
        )
        .await
        .map_err(map_database_error)?;
    let values: BTreeMap<String, String> = rows
        .into_iter()
        .map(|row| (row.config_key, row.config_value))
        .collect();
    Ok(json!({
        "systemName": required_value(&values, "system.name")?,
        "companyFullName": values.get("system.company.fullName").cloned().unwrap_or_default(),
        "logoUrl": values.get("system.logo").cloned().unwrap_or_default(),
        "watermarkEnabled": values.get("system.watermark.enabled").cloned().unwrap_or_else(|| "false".to_string()),
        "silentPrintEnabled": values.get("print.silent.enabled").cloned().unwrap_or_else(|| "false".to_string()),
    }))
}

pub async fn list(
    database: Option<&Arc<RBatis>>,
    query: &ConfigQuery,
) -> Result<PageResult<SysConfig>, AppError> {
    let database = require_database(database)?;
    let params = query.normalize_page()?;
    let offset = ((params.page - 1) * params.page_size) as i64;
    let mut where_sql = String::from(" where deleted = 0");
    let mut args = Vec::new();
    if let Some(config_name) = non_blank(query.config_name.as_deref()) {
        where_sql.push_str(" and config_name like ?");
        args.push(rbs::value!(format!("%{config_name}%")));
    }
    if let Some(config_key) = non_blank(query.config_key.as_deref()) {
        where_sql.push_str(" and config_key like ?");
        args.push(rbs::value!(format!("%{config_key}%")));
    }
    if let Some(status) = query.status {
        where_sql.push_str(" and status = ?");
        args.push(rbs::value!(status));
    }

    let mut page_args = args.clone();
    page_args.push(rbs::value!(params.page_size));
    page_args.push(rbs::value!(offset));
    let rows: Vec<SysConfig> = database
        .exec_decode(
            &format!(
                "select id, config_name, config_key, config_value, value_type, builtin, status, remark, deleted from sys_config {where_sql} order by created_at desc limit ? offset ?"
            ),
            page_args,
        )
        .await
        .map_err(map_database_error)?;
    let counts: Vec<CountRow> = database
        .exec_decode(
            &format!("select count(*) as total from sys_config {where_sql}"),
            args,
        )
        .await
        .map_err(map_database_error)?;
    Ok(PageResult {
        list: rows,
        total: I64String(counts.first().map(|row| row.total).unwrap_or(0)),
        page: params.page,
        page_size: params.page_size,
    })
}

pub async fn detail(database: Option<&Arc<RBatis>>, id: i64) -> Result<SysConfig, AppError> {
    let database = require_database(database)?;
    detail_with_database(database, id).await
}

pub async fn create(
    database: Option<&Arc<RBatis>>,
    request: ConfigSaveRequest,
) -> Result<I64String, AppError> {
    let database = require_database(database)?;
    validate_request(&request, true)?;
    let id = next_id();
    let value_type = type_or_default(request.value_type.as_deref()).to_string();
    database
        .exec(
            "insert into sys_config (id, config_name, config_key, config_value, value_type, builtin, status, remark, deleted) values (?, ?, ?, ?, ?, 0, ?, ?, 0)",
            vec![
                rbs::value!(id),
                rbs::value!(request.config_name),
                rbs::value!(request.config_key),
                rbs::value!(request.config_value.unwrap_or_default()),
                rbs::value!(value_type),
                rbs::value!(request.status.unwrap_or(1)),
                rbs::value!(request.remark),
            ],
        )
        .await
        .map_err(map_database_error)?;
    Ok(I64String(id))
}

pub async fn update(
    database: Option<&Arc<RBatis>>,
    id: i64,
    request: ConfigSaveRequest,
) -> Result<(), AppError> {
    let database = require_database(database)?;
    detail_with_database(database, id).await?;
    validate_request(&request, false)?;
    let value_type = type_or_default(request.value_type.as_deref()).to_string();
    database
        .exec(
            "update sys_config set config_name = ?, config_key = ?, config_value = ?, value_type = ?, status = ?, remark = ? where id = ? and deleted = 0",
            vec![
                rbs::value!(request.config_name),
                rbs::value!(request.config_key),
                rbs::value!(request.config_value.unwrap_or_default()),
                rbs::value!(value_type),
                rbs::value!(request.status.unwrap_or(1)),
                rbs::value!(request.remark),
                rbs::value!(id),
            ],
        )
        .await
        .map_err(map_database_error)?;
    Ok(())
}

pub async fn delete(database: Option<&Arc<RBatis>>, id: i64) -> Result<(), AppError> {
    let database = require_database(database)?;
    let current = detail_with_database(database, id).await?;
    if current.builtin == 1 {
        return Err(AppError::bad_request("operation failed"));
    }
    database
        .exec(
            "update sys_config set deleted = 1 where id = ? and deleted = 0",
            vec![rbs::value!(id)],
        )
        .await
        .map_err(map_database_error)?;
    Ok(())
}

pub async fn update_status(
    database: Option<&Arc<RBatis>>,
    id: i64,
    status: i32,
) -> Result<(), AppError> {
    let database = require_database(database)?;
    let current = detail_with_database(database, id).await?;
    if current.builtin == 1 {
        return Err(AppError::bad_request("operation failed"));
    }
    database
        .exec(
            "update sys_config set status = ? where id = ? and deleted = 0",
            vec![rbs::value!(status), rbs::value!(id)],
        )
        .await
        .map_err(map_database_error)?;
    Ok(())
}

async fn detail_with_database(database: &RBatis, id: i64) -> Result<SysConfig, AppError> {
    let rows: Vec<SysConfig> = database
        .exec_decode(
            "select id, config_name, config_key, config_value, value_type, builtin, status, remark, deleted from sys_config where id = ? and deleted = 0",
            vec![rbs::value!(id)],
        )
        .await
        .map_err(map_database_error)?;
    rows.into_iter()
        .next()
        .ok_or_else(|| AppError::not_found("资源不存在"))
}

fn require_database(database: Option<&Arc<RBatis>>) -> Result<&RBatis, AppError> {
    database
        .map(|database| database.as_ref())
        .ok_or_else(|| AppError::system("Rbatis database is not configured"))
}

fn required_value(values: &BTreeMap<String, String>, key: &str) -> Result<String, AppError> {
    let value = values.get(key).cloned().unwrap_or_default();
    if value.trim().is_empty() {
        return Err(AppError::system(format!("system config missing: {key}")));
    }
    Ok(value)
}

fn validate_request(request: &ConfigSaveRequest, require_value: bool) -> Result<(), AppError> {
    require_text(&request.config_name, "configName")?;
    require_text(&request.config_key, "configKey")?;
    if require_value && request.config_value.is_none() {
        return Err(AppError::bad_request("configValue is required"));
    }
    let value_type = type_or_default(request.value_type.as_deref());
    if let Some(value) = request.config_value.as_deref() {
        validate_value(value_type, value)?;
    }
    Ok(())
}

fn require_text(value: &str, field: &str) -> Result<(), AppError> {
    if value.trim().is_empty() {
        return Err(AppError::bad_request(format!("{field} is required")));
    }
    Ok(())
}

fn type_or_default(value_type: Option<&str>) -> &str {
    match value_type {
        Some(value) if !value.trim().is_empty() => value,
        _ => "string",
    }
}

fn validate_value(value_type: &str, value: &str) -> Result<(), AppError> {
    match value_type {
        "string" => Ok(()),
        "boolean" if value.eq_ignore_ascii_case("true") || value.eq_ignore_ascii_case("false") => {
            Ok(())
        }
        "boolean" => Err(AppError::bad_request("configValue must be true or false")),
        "number" => value
            .parse::<f64>()
            .map(|_| ())
            .map_err(|_| AppError::bad_request("configValue must be number")),
        _ => Err(AppError::bad_request(
            "valueType must be string, boolean or number",
        )),
    }
}

fn non_blank(value: Option<&str>) -> Option<&str> {
    value.map(str::trim).filter(|value| !value.is_empty())
}

fn map_database_error(err: rbatis::Error) -> AppError {
    let message = err.to_string();
    if message.to_ascii_lowercase().contains("duplicate") {
        AppError::conflict("数据冲突")
    } else {
        AppError::system(message)
    }
}
